google/cloud/internal/service_endpoint.cc

```diff
@@ -0,0 +1,49 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/internal/service_endpoint.h"
+#include "google/cloud/internal/absl_str_replace_quiet.h"
+#include "google/cloud/internal/make_status.h"
+#include "google/cloud/universe_domain_options.h"
+
+namespace google {
+namespace cloud {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace internal {
+
+// TODO(#13122): This function may need to be enhanced to support a env var
+// for UniverseDomain.
+StatusOr<std::string> DetermineServiceEndpoint(
+    absl::optional<std::string> endpoint_env_var,
+    absl::optional<std::string> endpoint_option, Options const& options,
+    std::string default_host) {
+  auto constexpr kGoogleDefaultUniverse = "googleapis.com";
+  if (endpoint_env_var.has_value()) return *endpoint_env_var;
+  if (endpoint_option.has_value()) return *endpoint_option;
```

Comment:
I think the plan is to call `DetermineServiceEndpoint` in the stub factory function. By that point the `EndpointOption` has already been defaulted, and will always have a value. So I think we would always break here.

https://github.com/googleapis/google-cloud-cpp/blob/4523a0a2b9ba461289c8f463ef0c1f65838e360d/google/cloud/accessapproval/v1/access_approval_connection.cc#L105-L106

Reply:
Can we modify `PopulateCommonOptions` to incorporate `UniverseDomainOption` ?

https://github.com/googleapis/google-cloud-cpp/blob/4523a0a2b9ba461289c8f463ef0c1f65838e360d/google/cloud/internal/populate_common_options.cc#L27